### PR TITLE
refactor: Replace WebMessageService with spring idiomatic ways to set response properties - part 2 [TECH-649]

### PR DESCRIPTION
**Pull request by jbee**
### Summary
This PR replaces use of `WebMessageService` with spring idiomatic ways of returning the controller result.

In essence this means instead of rendering the response to the HTTP response output stream it is returned from the method as `WebMessage` object. The rendering is done by jackson based on the `Accept` header.
This means some of the methods that did not explicitly declare a `produces` MIME type might now be more flexible based on the client's request.

Due to inconsistencies in some controller methods a number of non backwards compatible changes were made.
But these first take effect in 2.38.

### Automatic Testing
I added controller tests for each method that was changed.
Where setup wasn't to hard I even covered most of the paths in the methods (error and corner cases).
This means many many tests have been added.
However, these tests mostly are formal and do not check business logic. There mostly only check that the response code and message has the expected value and format since that might have been affected by the changes made.

### Backwards Compatibility
There are a number of methods which would render their report directly to the output in some of the code paths of the method but would return/render a full `WebMessage` in others. This is inconsistent and cannot be expressed in a form of a java return type.

Such methods now return a `WebMessage` for all code paths to be consistent with regards to the method itself but also to the overall DHIS2 API which usually returns a `WebMessage` envelope when a report of some sorts is returned.

However, such a change of always returning a `WebMessage` envelope around the report is a change of behaviour for those methods which had a code path that would not do that. To keep this behaviour at least prior to 2.38 I added a feature to the `WebMessage` to set a version prior to which the response is returned plain. When set (look for `withPlainResponseBefore`) the `response` property of the `WebMessage` is unpacked in the `WebMessageControllerAdvice` in case the requested API version means a plain response should be returned. This is covered by controller tests for all occurrences.

As the reponse sometimes uses a wrapper a slight change needed to be made to one wrapper to make sure the wrapper would have the same XML result as the unwrapped report has. This is only to keep prior 2.37 XML responses backwards compatible. 

### List of APIs that change in 2.38
For all endpoints listed below the API changes in a similar way. The response returned so far is wrapped in the `WebMessage` envelope. This means to access the response returned prior to 2.38 one has to extract the `response` object from the new root object. 

* `POST /completeDataSetRegistrations` with `JSON`/`XML` (only non `async` affected)
* `POST /dataValueSets` with `JSON`/`XML`/`ADX`/`CSV` (only non `async` affected)
* `POST /metadata` with `JSON`/`XML`/`GML`/`CSV` (only non `async` affected)
* `POST /predictions` (only non `async` affected)
* `PUT /predictions` (only non `async` affected)
* `PUT /relationships/{id}`
* `PUT /users/{uid}` with `JSON`/`XML`

Note that plain reports most often were returned with `200 OK` even if there were conflicts and the report had status `ERROR`.
Such responses will have HTTP response status `409 Conflict` when they are wrapped in a `WebMessage` and conflicts/errors are present in the report.

### CODE REVIEW

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/notification/NotificationMap.java [hunk 1]
```diff
@@ -84,8 +84,12 @@
 
     public void add( JobConfiguration configuration, Notification notification )
     {
-        JobType jobType = configuration.getJobType();
         String jobId = configuration.getUid();
+        if ( jobId == null )
+        {
+            return;
+        }
```
Comment (jbee) [inline dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/notification/NotificationMap.java:91]: FYI: This is a possible NPE fix I found during testing.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AccountController.java [hunk 2]
```diff
@@ -190,37 +191,32 @@ private void handleRecoveryLock( String username )
...
+    public WebMessage restoreAccount( @RequestParam String token, @RequestParam String password )
     {
         String[] idAndRestoreToken = securityService.decodeEncodedTokens( token );
         String idToken = idAndRestoreToken[0];
-        String restoreToken = idAndRestoreToken[1];
 
         UserCredentials credentials = userService.getUserCredentialsByIdToken( idToken );
 
-        if ( credentials == null )
+        if ( credentials == null || idAndRestoreToken.length < 2 )
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/AccountController.java:202]: FYI: adding `|| idAndRestoreToken.length < 2` is a fix for index out of bounds exception when calling with illegal `token`.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CompleteDataSetRegistrationController.java [hunk 3]
```diff
@@ -195,42 +193,37 @@ public void getCompleteRegistrationsJson(
...
-                .saveCompleteDataSetRegistrationsXml( request.getInputStream(), importOptions );
-            summary.setImportOptions( importOptions );
-            renderService.toXml( response.getOutputStream(), summary );
+            return asyncImport( importOptions, ImportCompleteDataSetRegistrationsTask.FORMAT_XML, request, response );
         }
+        response.setContentType( CONTENT_TYPE_XML );
+        ImportSummary summary = registrationExchangeService
+            .saveCompleteDataSetRegistrationsXml( request.getInputStream(), importOptions );
+        summary.setImportOptions( importOptions );
+        return WebMessageUtils.importSummary( summary );
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CompleteDataSetRegistrationController.java:210]: Obs! This is a change in behaviour. The `renderService.toXml( response.getOutputStream(), summary )` would only return the bare summary while this wraps the summary in the `WebMessage` envelope. This is needed to return a consistent type of result.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CompleteDataSetRegistrationController.java [hunk 4]
```diff
@@ -195,42 +193,37 @@ public void getCompleteRegistrationsJson(
...
-            ImportSummary summary = registrationExchangeService
-                .saveCompleteDataSetRegistrationsJson( request.getInputStream(), importOptions );
-            summary.setImportOptions( importOptions );
-            renderService.toJson( response.getOutputStream(), summary );
+            return asyncImport( importOptions, ImportCompleteDataSetRegistrationsTask.FORMAT_JSON, request, response );
         }
+        ImportSummary summary = registrationExchangeService
+            .saveCompleteDataSetRegistrationsJson( request.getInputStream(), importOptions );
+        summary.setImportOptions( importOptions );
+        return WebMessageUtils.importSummary( summary );
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/CompleteDataSetRegistrationController.java:226]: Obs! See comment above.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java [hunk 5]
```diff
@@ -557,31 +562,27 @@ public void unexpireUser( @PathVariable( "uid" ) String uid )
...
         HttpServletResponse response )
         throws Exception
     {
         User parsed = renderService.fromXml( request.getInputStream(), getEntityClass() );
 
-        ImportReport importReport = updateUser( pvUid, parsed );
-
-        response.setContentType( "application/xml" );
-        renderService.toXml( response.getOutputStream(), importReport );
+        return WebMessageUtils.importReport( updateUser( pvUid, parsed ) );
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java:573]: OBS! Again this is a change in the response as the report is wrapped in the `WebMessage` envelope.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java [hunk 6]
```diff
@@ -557,31 +562,27 @@ public void unexpireUser( @PathVariable( "uid" ) String uid )
...
         HttpServletResponse response )
         throws Exception
     {
         User parsed = renderService.fromJson( request.getInputStream(), getEntityClass() );
 
-        ImportReport importReport = updateUser( pvUid, parsed );
-
-        response.setContentType( "application/json" );
-        renderService.toJson( response.getOutputStream(), importReport );
+        return WebMessageUtils.importReport( updateUser( pvUid, parsed ) );
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/user/UserController.java:585]: OBS! Again this is a change in the response as the report is wrapped in the `WebMessage` envelope.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/sms/SmsGatewayController.java [hunk 7]
```diff
@@ -114,62 +112,62 @@ public void getGatewayConfiguration( @PathVariable String uid, HttpServletRespon
...
         if ( config == null )
         {
-            throw new WebMessageException( WebMessageUtils.notFound( "No gateway found" ) );
+            return WebMessageUtils.notFound( "No gateway found" );
         }
 
         SmsGatewayConfig updatedConfig = renderService.fromJson( request.getInputStream(), SmsGatewayConfig.class );
 
-        if ( gatewayAdminService.hasDefaultGateway() && updatedConfig.isDefault() )
+        if ( gatewayAdminService.hasDefaultGateway() && updatedConfig.isDefault() && !config.isDefault() )
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/sms/SmsGatewayController.java:144]: Note that adding `&& !config.isDefault()` is a fix so that the default config can be updated.

File: dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/event/RelationshipControllerTest.java [hunk 8]
```diff
@@ -177,18 +176,11 @@ public void testGetRelationship()
         mockMvc.perform( get( ENDPOINT + "/" + REL_ID ) ).andExpect( status().isOk() );
     }
 
-    @Test( expected = NestedServletException.class )
-    public void testDeleteRelationshipNotPresent()
-        throws Exception
-    {
-        mockMvc.perform( delete( ENDPOINT + "/" + REL_ID ) ).andExpect( status().isConflict() );
-    }
-
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/test/java/org/hisp/dhis/webapi/controller/event/RelationshipControllerTest.java:186]: FYI: I removed this test since it now tested via actual controller interaction and would here assert `OK` which is confusing. The reason this test would need to assert `OK` when the actual result is `NOT_FOUND` is that the change to the HTTP response status is first made in an controller advice which is outside of this test setup.

**Review by netroms — APPROVED** [review]

**Review by mortenoh — APPROVED** [review]
